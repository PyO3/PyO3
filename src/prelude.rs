//! PyO3's prelude.
//!
//! The purpose of this module is to alleviate imports of many commonly used items of the PyO3 crate
//! by adding a glob import to the top of pyo3 heavy modules:
//!
//! ```
//! # #![allow(unused_imports)]
//! use pyo3::prelude::*;
//! ```

pub use crate::conversion::{FromPyObject, IntoPyObject};
#[allow(deprecated)]
pub use crate::conversion::{IntoPy, ToPyObject};
pub use crate::err::{PyErr, PyResult};
pub use crate::instance::{Borrowed, Bound, Py, PyObject};
pub use crate::marker::Python;
pub use crate::pycell::{PyRef, PyRefMut};
pub use crate::pyclass_init::PyClassInitializer;
pub use crate::types::{PyAny, PyModule};

#[cfg(feature = "macros")]
pub use pyo3_macros::{
    pyclass, pyfunction, pymethods, pymodule, FromPyObject, IntoPyObject, IntoPyObjectRef,
};

#[cfg(feature = "macros")]
pub use crate::wrap_pyfunction;
#[cfg(feature = "macros")]
#[allow(deprecated)]
pub use crate::wrap_pyfunction_bound;

pub use crate::types::any::PyAnyMethods;
pub use crate::types::weakref::PyWeakrefMethods;
