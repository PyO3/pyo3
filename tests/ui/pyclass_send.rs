use pyo3::prelude::*;
use std::os::raw::c_void;

#[pyclass]
struct NotSyncNotSend(*mut c_void);

#[pyclass]
struct SendNotSync(*mut c_void);
unsafe impl Send for SendNotSync {}

#[pyclass]
struct SyncNotSend(*mut c_void);
unsafe impl Sync for SyncNotSend {}

// None of the `unsendable` forms below should fail to compile

#[pyclass(unsendable)]
struct NotSyncNotSendUnsendable(*mut c_void);

#[pyclass(unsendable)]
struct SendNotSyncUnsendable(*mut c_void);
unsafe impl Send for SendNotSyncUnsendable {}

#[pyclass(unsendable)]
struct SyncNotSendUnsendable(*mut c_void);
unsafe impl Sync for SyncNotSendUnsendable {}

fn main() {}
