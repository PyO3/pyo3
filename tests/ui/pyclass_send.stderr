error[E0277]: `*mut c_void` cannot be shared between threads safely
 --> tests/ui/pyclass_send.rs:5:8
  |
5 | struct NotSyncNotSend(*mut c_void);
  |        ^^^^^^^^^^^^^^ `*mut c_void` cannot be shared between threads safely
  |
  = help: within `NotSyncNotSend`, the trait `Sync` is not implemented for `*mut c_void`
note: required because it appears within the type `NotSyncNotSend`
 --> tests/ui/pyclass_send.rs:5:8
  |
5 | struct NotSyncNotSend(*mut c_void);
  |        ^^^^^^^^^^^^^^
note: required by a bound in `assert_pyclass_sync`
 --> src/impl_/pyclass/assertions.rs
  |
  | pub const fn assert_pyclass_sync<T>()
  |              ------------------- required by a bound in this function
  | where
  |     T: PyClassSync + Sync,
  |                      ^^^^ required by this bound in `assert_pyclass_sync`

error[E0277]: `*mut c_void` cannot be sent between threads safely
 --> tests/ui/pyclass_send.rs:4:1
  |
4 | #[pyclass]
  | ^^^^^^^^^^ `*mut c_void` cannot be sent between threads safely
  |
  = help: within `NotSyncNotSend`, the trait `Send` is not implemented for `*mut c_void`
  = help: the trait `pyo3::impl_::pyclass::PyClassThreadChecker<T>` is implemented for `SendablePyClass<T>`
note: required because it appears within the type `NotSyncNotSend`
 --> tests/ui/pyclass_send.rs:5:8
  |
5 | struct NotSyncNotSend(*mut c_void);
  |        ^^^^^^^^^^^^^^
  = note: required for `SendablePyClass<NotSyncNotSend>` to implement `pyo3::impl_::pyclass::PyClassThreadChecker<NotSyncNotSend>`
note: required by a bound in `PyClassImpl::ThreadChecker`
 --> src/impl_/pyclass.rs
  |
  |     type ThreadChecker: PyClassThreadChecker<Self>;
  |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `PyClassImpl::ThreadChecker`
  = note: this error originates in the attribute macro `pyclass` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: `*mut c_void` cannot be shared between threads safely
 --> tests/ui/pyclass_send.rs:8:8
  |
8 | struct SendNotSync(*mut c_void);
  |        ^^^^^^^^^^^ `*mut c_void` cannot be shared between threads safely
  |
  = help: within `SendNotSync`, the trait `Sync` is not implemented for `*mut c_void`
note: required because it appears within the type `SendNotSync`
 --> tests/ui/pyclass_send.rs:8:8
  |
8 | struct SendNotSync(*mut c_void);
  |        ^^^^^^^^^^^
note: required by a bound in `assert_pyclass_sync`
 --> src/impl_/pyclass/assertions.rs
  |
  | pub const fn assert_pyclass_sync<T>()
  |              ------------------- required by a bound in this function
  | where
  |     T: PyClassSync + Sync,
  |                      ^^^^ required by this bound in `assert_pyclass_sync`

error[E0277]: `*mut c_void` cannot be sent between threads safely
  --> tests/ui/pyclass_send.rs:11:1
   |
11 | #[pyclass]
   | ^^^^^^^^^^ `*mut c_void` cannot be sent between threads safely
   |
   = help: within `SyncNotSend`, the trait `Send` is not implemented for `*mut c_void`
   = help: the trait `pyo3::impl_::pyclass::PyClassThreadChecker<T>` is implemented for `SendablePyClass<T>`
note: required because it appears within the type `SyncNotSend`
  --> tests/ui/pyclass_send.rs:12:8
   |
12 | struct SyncNotSend(*mut c_void);
   |        ^^^^^^^^^^^
   = note: required for `SendablePyClass<SyncNotSend>` to implement `pyo3::impl_::pyclass::PyClassThreadChecker<SyncNotSend>`
note: required by a bound in `PyClassImpl::ThreadChecker`
  --> src/impl_/pyclass.rs
   |
   |     type ThreadChecker: PyClassThreadChecker<Self>;
   |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `PyClassImpl::ThreadChecker`
   = note: this error originates in the attribute macro `pyclass` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: `*mut c_void` cannot be sent between threads safely
 --> tests/ui/pyclass_send.rs:4:1
  |
4 | #[pyclass]
  | ^^^^^^^^^^ `*mut c_void` cannot be sent between threads safely
  |
  = help: within `NotSyncNotSend`, the trait `Send` is not implemented for `*mut c_void`
note: required because it appears within the type `NotSyncNotSend`
 --> tests/ui/pyclass_send.rs:5:8
  |
5 | struct NotSyncNotSend(*mut c_void);
  |        ^^^^^^^^^^^^^^
note: required by a bound in `SendablePyClass`
 --> src/impl_/pyclass.rs
  |
  | pub struct SendablePyClass<T: Send>(PhantomData<T>);
  |                               ^^^^ required by this bound in `SendablePyClass`
  = note: this error originates in the attribute macro `pyclass` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: `*mut c_void` cannot be sent between threads safely
  --> tests/ui/pyclass_send.rs:11:1
   |
11 | #[pyclass]
   | ^^^^^^^^^^ `*mut c_void` cannot be sent between threads safely
   |
   = help: within `SyncNotSend`, the trait `Send` is not implemented for `*mut c_void`
note: required because it appears within the type `SyncNotSend`
  --> tests/ui/pyclass_send.rs:12:8
   |
12 | struct SyncNotSend(*mut c_void);
   |        ^^^^^^^^^^^
note: required by a bound in `SendablePyClass`
  --> src/impl_/pyclass.rs
   |
   | pub struct SendablePyClass<T: Send>(PhantomData<T>);
   |                               ^^^^ required by this bound in `SendablePyClass`
   = note: this error originates in the attribute macro `pyclass` (in Nightly builds, run with -Z macro-backtrace for more info)
